File: 121.best-time-to-buy-and-sell-stock.cpp
// @leet start
#include <vector>
using std::vector;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int best = 0;
        int lo = 0;
        int hi = 1;
        int n = prices.size();
        
        while (hi < n) {
            if (prices[hi] <= prices[lo]) {
                lo = hi;
            }
            best = std::max(prices[hi] - prices[lo], best);
            ++hi;
        }

        return best;
    }
};
// @leet end

File: 121.best-time-to-buy-and-sell-stock.py
from typing import List


# @leet start
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        ans = 0
        lowest = prices[0]
        n = len(prices)

        for i in range(n):
            ans = max(ans, prices[i] - lowest)
            lowest = min(lowest, prices[i])

        return ans


# @leet end
